src/core/Ekubo.sol

```diff
@@ -0,0 +1,453 @@
+                // `settle`'d. `globalSellAmount` is the verbatim credit in that token stored by the
+                // vault. We only need to handle the case of incomplete filling.
+                if (globalSellAmount != 0) {
+                    Take._callSelector(
+                        uint32(IEkuboCore.withdraw.selector),
+                        globalSellToken,
+                        (payer == address(this))? address(this) : _msgSender(),
+                        globalSellAmount
+                    );
+                }
+            } else {
+                // While `notes` records a credit value, the vault actually records a debt for the
+                // global sell token. We recover the exact amount of that debt and then pay it.
+                // `globalSellAmount` is _usually_ zero, but if it isn't it represents a partial
+                // fill. This subtraction recovers the actual debt recorded in the vault.
+                uint256 debt;
+                unchecked {
+                    debt = state.globalSellAmount - globalSellAmount;
+                }
+                if (debt == 0) {
+                    revert ZeroSellAmount(globalSellToken);
+                }
+                _ekuboPay(globalSellToken, payer, debt, permit, isForwarded, sig);
+            }
+
+            // return abi.encode(globalBuyAmount);
+            bytes memory returndata;
+            assembly ("memory-safe") {
+                returndata := mload(0x40)
+                mstore(returndata, 0x60)
+                mstore(add(0x20, returndata), 0x20)
+                mstore(add(0x40, returndata), 0x20)
+                mstore(add(0x60, returndata), globalBuyAmount)
+                mstore(0x40, add(0x80, returndata))
+            }
+            return returndata;
+        }
+    }
+
+    function payCallback(bytes calldata data) private returns (bytes memory) {
+        IERC20 sellToken;
+        uint256 requestedAmount;
+        uint256 sellAmount;
+        
+        ISignatureTransfer.PermitTransferFrom calldata permit;
+        bool isForwarded;
+        bytes calldata sig;
+
+        
+        assembly ("memory-safe") {
+            
+        }
+        
+        assembly ("memory-safe") {
+            // Initialize permit and sig to appease the compiler
+            permit := calldatasize()
+            sig.offset := calldatasize()
+            sig.length := 0x00
+
+            sellToken := calldataload(data.offset)
+            requestedAmount := calldataload(add(0x20, data.offset))
+            sellAmount := calldataload(add(0x40, data.offset))
+
+            if gt(calldatasize(), 0x60) {
+                permit := add(0x60, data.offset)
+                isForwarded := calldataload(add(0xc0, data.offset))
+                sig.length := calldataload(add(0xe0, data.offset))
+                sig.offset := add(0x100, data.offset)
+            }
+        }
+        // TODO: Check if  assert(requestedAmount == sellAmount) should be true
```

Comment:
This is almost certainly not the case. With this kind of AMM `requestedAmount <= sellAmount` because the AMM may run out of liquidity.

Reply:
Thought so as well, will make sure of using requestedAmount